### PR TITLE
fix(robotwin): support pi0 eval on robotwin

### PR DESCRIPTION
**Pull request by Iron-Wph**
### Description

1. Add the "center_crop" configuration to support both OpenVLA-OFT and PI0;
2. Modify the configurations of the optimizer and learning rate scheduler to adapt to the robotwin environment, see robotwin_sft_openpi.yaml;
3. support different data's key for SFT and RL for pi0 on RoboTwin;
4. Support PI0 evaluation on the robotwin environment.

### Additional information (optional, e.g., figures and logs):
1. For PI0 evaluation, it is recommended to use the embodiment [aloha-agilex], instead of the embodiment [piper, piper, 0.6] which is applied for OpenVLA-OFT
2. Eval result on adjust_bottle_clean task: 
PixPin_2026-02-15_13-30-55

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/embodiment/openpi/openpi_action_model.py [hunk 1]
```diff
@@ -294,23 +294,36 @@ def default_forward(
...
-            processed_obs["observation/state_gripper"] = state[:, 6:7]
-        else:
-            processed_obs["observation/state"] = env_obs["states"]
-        # wrist image observation
-        if env_obs["wrist_images"] is not None:
-            processed_obs["observation/wrist_image"] = env_obs["wrist_images"]
-        # store used keys
+        if "robotwin" in self.config.config_name:
+            processed_obs = {
+                "images": {},
```
Comment (XuS1994) [inline rlinf/models/embodiment/openpi/openpi_action_model.py]: `key` need to be kept consistent to reduce the cost of debug in the future.

images --> observation/image
state --> observation/state

File: rlinf/models/embodiment/openpi/openpi_action_model.py [hunk 2]
```diff
@@ -294,23 +294,36 @@ def default_forward(
...
+                "prompt": env_obs["task_descriptions"],
+                'state': env_obs["states"],
+            }
+            # [b, h, w, c] -> [b, c, h, w]           
+            processed_obs["images"]["cam_high"] = env_obs["main_images"].permute(0,3,1,2).contiguous()
+            if env_obs["wrist_images"] is not None:
+                processed_obs["images"]["cam_left_wrist"] = env_obs["wrist_images"].permute(0,3,1,2).contiguous()
+            if env_obs["extra_view_images"] is not None:
+                processed_obs["images"]["cam_right_wrist"] = env_obs["extra_view_images"].permute(0,3,1,2).contiguous()
+        else:
```
Comment (XuS1994) [inline rlinf/models/embodiment/openpi/openpi_action_model.py]: better to set `observation/image` `prompt` `states` individually.
like this:
```
processed_obs = {
            "prompt": env_obs["task_descriptions"],
}
 
# for image
if "robotwin" in self.config.config_name:
            processed_obs["observation/image"] = xxx
else:
            processed_obs["observation/image"] = env_obs["main_images"]

# for state
if "calvin" in self.config.config_name:
                state = env_obs["states"]
                processed_obs["observation/state_ee_pos"] = state[:, :3]
                processed_obs["observation/state_ee_rot"] = state[:, 3:6]
                processed_obs["observation/state_gripper"] = state[:, 6:7]
else:
                processed_obs["observation/state"] = env_obs["states"]
```

Comment (Iron-Wph) [reply rlinf/models/embodiment/openpi/openpi_action_model.py]: I have modified the code to ensure consistent data keys are used.

File: rlinf/models/embodiment/openpi/policies/aloha_policy.py [hunk 3]
```diff
@@ -178,25 +178,37 @@ def _gripper_from_angular_inv(value):
...
+    if img is None:
+        return None
+    img = np.asarray(img)
+    # Convert to uint8 if using float images.
+    if np.issubdtype(img.dtype, np.floating):
+        img = (255 * img).astype(np.uint8)
+    if img.shape[0] == 3:
+        img = einops.rearrange(img, "c h w -> h w c")
+    elif img.shape[0] == 2 and img.shape[1] == 3:
+        img = einops.rearrange(img, "n c h w -> n h w c")
```
Comment (QuanluZhang) [inline rlinf/models/embodiment/openpi/policies/aloha_policy.py]: missing error handling. what if img.shape[0] is neither 3 nor (2 and shape[1]==3)? Function returns image without rearranging, which might have wrong dimensions.

the following is a suggestion from coding agent, please check whether it is correct and better, based on your own judgment.
```
    if img.shape[0] == 3 and img.shape[1] != 3:  # CHW format
        img = einops.rearrange(img, "c h w -> h w c")
    elif img.ndim == 4 and img.shape[1] == 3:  # NCHW format
        img = einops.rearrange(img, "n c h w -> n h w c")
    elif img.shape[-1] != 3:  # Neither HWC nor NHWC
        raise ValueError(f"Unexpected image shape: {img.shape}")
    return img
```

File: rlinf/models/embodiment/openpi/dataconfig/__init__.py [hunk 4]
```diff
@@ -218,16 +218,19 @@
...
-        model=pi0_config.Pi0Config(),
+        model=pi0_config.Pi0Config(discrete_state_input=False),
         data=LeRobotAlohaDataConfig(
             repo_id="robotwin/place_empty_cup_random",
+            adapt_to_pi=False,
             base_config=DataConfig(
                 prompt_from_task=True
             ),  # we need language instruction
-            assets=AssetsConfig(assets_dir="checkpoints/torch/pi0_robotwin/assets"),
+            assets=AssetsConfig(assets_dir="checkpoints/torch/pi0_aloha_robotwin/assets"),
```
Comment (QuanluZhang) [inline rlinf/models/embodiment/openpi/dataconfig/__init__.py]: why do you change the assets path? can the previously downloaded assets still be used correctly, or I should download the new assets?

File: rlinf/envs/robotwin/robotwin_env.py [hunk 5]
```diff
@@ -170,12 +172,12 @@ def _extract_obs_image(self, raw_obs):
         batch_states = []
         batch_instructions = []
         for obs in raw_obs:
-            batch_images.append(self.center_and_crop(obs["full_image"]))
+            batch_images.append(self.center_and_crop(obs["full_image"], center_crop=self.center_crop))
```
Comment (QuanluZhang) [inline rlinf/envs/robotwin/robotwin_env.py]: remove the trailing spaces

File: rlinf/models/embodiment/openpi/policies/aloha_policy.py [hunk 6]
```diff
@@ -178,25 +178,37 @@ def _gripper_from_angular_inv(value):
...
+    return img
 
 def _decode_aloha(data: dict, *, adapt_to_pi: bool = False) -> dict:
     # state is [left_arm_joint_angles, left_arm_gripper, right_arm_joint_angles, right_arm_gripper]
-    # dim sizes: [6, 1, 6, 1]
-    state = np.asarray(data["state"])
+    # dim sizes: [6, 1, 6, 1]    
+    if 'observation/state' in data:
```
Comment (QuanluZhang) [inline rlinf/models/embodiment/openpi/policies/aloha_policy.py]: please use " instead of ' to be consistent with the code style

File: rlinf/models/embodiment/openpi/dataconfig/__init__.py (hunk 4 above)
Comment (Iron-Wph) [reply rlinf/models/embodiment/openpi/dataconfig/__init__.py]: This is only to match the configuration name. In practice, if you intend to perform SFT (Supervised Fine-Tuning) on RoboTwin with the pi0 model, you need to execute the following commands. For more details, refer to: https://robotwin-platform.github.io/doc/usage/Pi0.html#5-5-finetune-model:
```
uv run scripts/compute_norm_stats.py --config-name ${train_config_name} 
cp -r your_path/to/assets/your_config_name your_path/to/pi0_base
```

File: rlinf/models/embodiment/openpi/policies/aloha_policy.py (hunk 6 above)
Comment (Iron-Wph) [reply rlinf/models/embodiment/openpi/policies/aloha_policy.py]: The reason for this is that the data key naming differs between SFT and evaluation processes (note that evaluation adheres to the same RL training logic). Specifically:
In SFT, the 'state' value is accessible via data['state'];
In evaluation, 'state' is stored under data['observation/state'].
This naming rule must be strictly followed, as specified in the code file RLinf/rlinf/models/embodiment/openpi/openpi_action_model.py (line 296, obs_processor function).

File: rlinf/models/embodiment/openpi/policies/aloha_policy.py (hunk 3 above)
Comment (Iron-Wph) [reply rlinf/models/embodiment/openpi/policies/aloha_policy.py]: Just confirmed: the image shape is correct, so no additional processing is needed.

File: rlinf/envs/robotwin/robotwin_env.py (hunk 5 above)
Comment (Iron-Wph) [reply rlinf/envs/robotwin/robotwin_env.py]: haved solved

File: rlinf/models/embodiment/openpi/dataconfig/__init__.py (hunk 4 above)
Comment (QuanluZhang) [reply rlinf/models/embodiment/openpi/dataconfig/__init__.py]: @guozhen1997 1) is the path "checkpoints/torch/pi0_aloha_robotwin/assets" used in RLinf? 2) do we mention the manual copy in our doc (i.e., `cp -r your_path/to/assets/your_config_name your_path/to/pi0_base`)?

**Review by QuanluZhang — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]
